apps/src/lib/node/ledger/shell/process_proposal.rs

```diff
@@ -34,16 +35,31 @@ where
         &mut self,
         req: RequestProcessProposal,
     ) -> ResponseProcessProposal {
+        // the number of vote extension digests included in the block proposal
+        let mut vote_ext_digest_num = 0;
         let tx_results: Vec<ExecTxResult> = req
             .txs
             .iter()
             .map(|tx_bytes| {
-                ExecTxResult::from(self.process_single_tx(tx_bytes))
+                ExecTxResult::from(
+                    self.process_single_tx(tx_bytes, &mut vote_ext_digest_num),
+                )
             })
             .collect();
 
+        // We should not have more than one `VoteExtensionDigest` in
+        // a proposal from some round's leader.
+        let too_many_vext_digests = vote_ext_digest_num > 1;
+
+        // Erroneous transactions were detected when processing
+        // the leader's proposal. We allow txs that do not
+        // deserialize properly, that have invalid signatures
+        // and that have invalid wasm code to reach FinalizeBlock.
+        // This is the reason behind that greater than three check.
+        let invalid_txs = tx_results.iter().any(|res| res.code > 3);
```

Comment:
Maybe we make a constant `PROTOCOL_LEVEL_ERROR=3`. We allow user txs to go through that error, but if something is wrong at the protocol level, we reject. That might make this code a little clearer to understand

Reply:
And if you think of a better name than `PROTOCOL_LEVEL_ERROR`, use it.